api/common/sharing/utils.go

```diff
@@ -12,44 +12,234 @@ import (
+	if IDExists && len(ID.String()) > 0 && hasAtLeastOneSelector {
+		return util.HandleInstanceSharingError(util.ErrInvalidReferenceSelectors, "")
+	}
+	return nil
+}
 
-	return referencedInstanceID.String(), nil
+func getSharedInstances(ctx context.Context, repository storage.Repository, offeringID, tenantIdentifier, tenantID string) (types.ObjectList, error) {
+	queryParams := map[string]interface{}{
+		"tenant_identifier": tenantIdentifier,
+		"tenant_id":         tenantID,
+		"offering_id":       offeringID,
+	}
+
+	objectList, err := repository.QueryForList(ctx, types.ServiceInstanceType, storage.QueryForSharedInstances, queryParams)
+	return objectList, err
+}
+
+func filterInstancesBySelectors(ctx context.Context, repository storage.Repository, instances types.ObjectList, parameters map[string]gjson.Result, smaap bool, offeringID string) (types.ServiceInstances, error) {
+	var err error
+	var filteredInstances types.ServiceInstances
+
+	planSelectorVal, planSelectorExists := parameters[instance_sharing.ReferencePlanNameSelector]
+	var selectorPlan *types.ServicePlan
+	if planSelectorExists {
+		selectorPlan, err = getSelectorPlan(ctx, repository, smaap, offeringID, planSelectorVal.String())
+		if err != nil {
+			log.C(ctx).Errorf("Failed to retrieve the plan %s with the offering %s via the selector provided: %s", planSelectorVal.String(), offeringID, err)
+			return types.ServiceInstances{}, err
+		}
+	}
+
+	for i := 0; i < instances.Len(); i++ {
+
+		// "*" for any shared instance of the tenant, if parameter not passed, set true for any shared instance
+		selectorVal, exists := parameters[instance_sharing.ReferencedInstanceIDKey]
+		if exists && selectorVal.String() != "*" {
+			continue
+		}
+
+		instance := instances.ItemAt(i).(*types.ServiceInstance)
+
+		selectorVal, exists = parameters[instance_sharing.ReferenceInstanceNameSelector]
+		if exists && len(selectorVal.String()) > 0 && !(instance.Name == selectorVal.String()) {
+			continue
+		}
+
+		if planSelectorExists && (selectorPlan == nil || instance.ServicePlanID != selectorPlan.ID) {
+			continue
+		}
+
+		selectorVal, exists = parameters[instance_sharing.ReferenceLabelSelector]
+		if exists && len(selectorVal.String()) > 0 {
+			match, err := matchLabels(instance.Labels, []byte(selectorVal.Raw))
+			if err != nil {
+				return types.ServiceInstances{}, err
+			}
+			if !match {
+				continue
+			}
+		}
+
+		filteredInstances.Add(instance)
+		// only single result is accepted for selectors:
+		if filteredInstances.Len() > 1 {
+			log.C(ctx).Errorf("%s", util.ErrMultipleReferenceSelectorResults)
+			return types.ServiceInstances{}, util.HandleInstanceSharingError(util.ErrMultipleReferenceSelectorResults, "")
+		}
+	}
+	if filteredInstances.Len() == 0 {
+		log.C(ctx).Errorf("%s", util.ErrNoResultsForReferenceSelector)
+		return types.ServiceInstances{}, util.HandleInstanceSharingError(util.ErrNoResultsForReferenceSelector, "")
+	}
+	return filteredInstances, nil
+}
```

Comment:
return *types.ServiceInstance no need for array

Reply:
done